### ISSUE TITLE
Flush-before-exit does not apply to Stalker

### ISSUE THREAD

**Issue by Ayrx** [frida/frida-core#218]
Frida flushes pending data before a program exits.

> Frida will now intercept the most common process termination APIs and take care of flushing any pending data for you.

This currently does not apply to the data in Stalker's queue. This should be added so the data in Stalker isn't lost.

**Comment by Ayrx** [comment]
I'll see if I can implement this when I have some spare time to work on it.

**Comment by Ayrx** [comment]
Hi @oleavr,

So, my thoughts are that the first step should be adding a `Stalker.drainQueue()` method that allows someone to drain the queue whenever they want to.

I'm a little confused by the architecture of `frida-gum` so I hope you can help me out here. :)

First, it looks like I need to add to `gumdukstalker.c` the following to declare a new JS function?:

```c
GUMJS_DECLARE_FUNCTION (gumjs_stalker_drain)

GUMJS_DEFINE_FUNCTION (gumjs_stalker_drain)
{
  ...
}
```

It appears that I also need to add an entry to `static const duk_function_list_entry gumjs_stalker_functions[]`? But I can't figure out what the number that is the 3rd element in the array is meant to be.

Next, I'm also not sure how to actually implement the `gumjs_stalker_drain` function.

From what I can tell, the `gum_duk_event_sink_drain` function in `gumdukeventsink.c` is responsible for actually draining the queue? I suppose I have to call `gum_duk_event_sink_drain` from my `gumjs_stalker_drain` function? But how do I actually obtain a pointer to the `GumDukEventSink` object in use?

And finally I'll have to repeat the same thing for the v8 implementation?

Thanks in advance for the assist!

**Comment by oleavr** [comment]
> So, my thoughts are that the first step should be adding a Stalker.drainQueue() method that allows someone to drain the queue whenever they want to.

Sounds good. Let's name it `Stalker.flush()` though, for consistency with `Interceptor.flush()`.

> It appears that I also need to add an entry to static const duk_function_list_entry gumjs_stalker_functions[]? But I can't figure out what the number that is the 3rd element in the array is meant to be.

It's the maximum number of arguments, so it should be `0` in this case.

> From what I can tell, the gum_duk_event_sink_drain function in gumdukeventsink.c is responsible for actually draining the queue?

Correct. It's run on Frida's JS thread, but because it grabs the JS lock it is safe to call it from other threads too.

> I suppose I have to call gum_duk_event_sink_drain from my gumjs_stalker_drain function?

Correct.

> But how do I actually obtain a pointer to the GumDukEventSink object in use?

Like in some of the other methods you have to get a pointer to the module itself:

```C
  GumDukStalker * module;

  module = gumjs_module_from_args (args);
```

And we need to store the sinks by adding a field [here](https://github.com/frida/frida-gum/blob/a1a088e3f6ca13691f01cc93fb6d7636c31a19b9/bindings/gumjs/gumdukstalker.h#L28). The current logic hands the sink to the Stalker, so we need to change that so we keep e.g. a [HashTable](https://developer.gnome.org/glib/unstable/glib-Hash-Tables.html) used as a set, i.e. [g_hash_table_add()](https://developer.gnome.org/glib/unstable/glib-Hash-Tables.html#g-hash-table-add) and [g_hash_table_remove()](https://developer.gnome.org/glib/unstable/glib-Hash-Tables.html#g-hash-table-remove). That won't need an explicit lock since we're already holding the JS lock.

> And finally I'll have to repeat the same thing for the v8 implementation?

Correct.

**Comment by oleavr** [comment]
Actually, we should solve this at the Stalker level instead. The reason is that we also need the agent, in frida-core/lib/agent, to be able to flush as part of its exit-handling. GumJS is only a language binding for Gum, so we should avoid putting such logic there.

So we basically need to add:
```C
GUM_API void gum_stalker_flush (GumStalker * self);
```
Right above `gum_stalker_stop()` in `gum/gumstalker.h`.

This should grab the Stalker lock and iterate all of the contexts:

```C
  GSList * cur;

  GUM_STALKER_LOCK (self);

  for (cur = self->contexts; cur != NULL; cur = cur->next)
  {
    GumExecCtx * ctx = (GumExecCtx *) cur->data;
    ...
  }

  GUM_STALKER_UNLOCK (self);
```

And that should just collect all the sinks, taking a reference on each. Something like:

```C
  GSList * sinks, * cur;

  GUM_STALKER_LOCK (self);

  sinks = NULL;
  for (cur = self->contexts; cur != NULL; cur = cur->next)
  {
    GumExecCtx * ctx = cur->data;
    sinks = g_slist_prepend (sinks, g_object_ref (ctx->sink));
  }

  GUM_STALKER_UNLOCK (self);
```

After letting go of the lock, after the loop, it should iterate the collection of sinks and call `flush()` on each:

```C
  for (cur = sinks; cur != NULL; cur = cur->next)
  {
    GumEventSink * sink = cur->data;
    gum_event_sink_flush (sink);
  }
```

And then release the sinks along with the list itself:

```C
  g_slist_free_full (sinks, g_object_unref);
```

We also need to extend the `EventSink` interface (in `gum/gumeventsink.h`) to add a `flush` method right above `stop`. This should just be a matter of copy-pasting the `stop` occurrences and renaming.

**Comment by Ayrx** [comment]
This has been fixed in Frida 12.2.21. :)
